jaxrs/WFLY-13530_expose_JAXRS_to_gRPC.adoc

```diff
@@ -0,0 +1,133 @@
+= Expose JAX-RS Resources to gRPC clients
+:author:            Ron Sigal
+:email:             rsigal@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org/browse
+
+== Overview
+
+In the five years since gRPC was introduced, it has become a popular invocation
+and streaming framework, used by major corporations. Among its benefits are
+
+* speed
+* language independence
+
+With this feature request we propose to expose JAX-RS services as gRPC services,
+accessible by gRPC clients.
+
+Given a gRPC service definition like
+----
+service Greeter {
+  rpc SayHello (HelloRequest) returns (HelloReply) {}
+}
+----
+
+the protobuf java compiler will create a class GreeterGrpc, and a service can be
+created by subclassing. We envision making it as easy as something like this
+to forward a gRPC request to JAX-RS:
+
+----
+
+static class GreeterImpl extends GreeterGrpc.GreeterImplBase {
+
+   public void sayHello(HelloRequest req, StreamObserver<HelloReply> responseObserver) {
+      JAXRSForwarderBuilder builder = new JAXRSForwarderBuilder();
+      builder.servlet("ResteasyServlet").pathTranslator((String s) -> ("test/" + s));
+      JAXRSForwarder forwarder = builder.build();
+      forwarder.forward(req, (StreamObserver) responseObserver);
+   }
+}
+----
+
+The JAXRSForwarder will find the appropriate servlet, create the appropriate HTTP
+requests and responses, call the JAX-RS resource method, and send
+the result to the gRPC StreamObserver. 
+
+== Issue Metadata
+
+=== Issue
+
+* {issue-base-url}/EAP7-1536 [EAP7-1536]
+* {issue-base-url}/WFLY-13530 [WFLY-13530]
+
+
+=== Related Issues
+
+* {issue-base-url}/PRODMGT-1811 [PRODMGT-1811]
+* {issue-base-url}/WFLY-11034 [WFLY-11034]
+* {issue-base-url}/RESTEASY-2656 [RESTEASY-2656]
+
+=== Dev Contacts
+
+* mailto:rsigal@redhat.com[Ron Sigal]
+* mailto:asoldano@redhat.com[Alessio Soldano]
+* mailto:ema@redhat.com[Jim Ma]
+* mailto:rsearls@redhat.com[Rebecca Searls]
+* mailto:weli@redhat.com[Wei Nan Li]
+* mailto:jperkins@redhat.com[James Perkins]
+
+=== QE Contacts
+
+* ???
+
+=== Testing By
+
+TBD
+
+* [ ] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+* RESTEasy
+* Wildfly
+
+=== Other Interested Projects
+
+* Elytron ?
```

Comment:
I think this will definitely be required so we need to make sure the security integration is a part of the requirements.  What I want to avoid is something where we end up with some temporary security solution where we retrospectively need to add Elytron integration and than have two solutions in play whilst one is deprecated.

It may be we can eliminate the client side for now although if invocations originating within a server are required we probably should not.

Both TLS and authentication do seem to be defined for gRPC so Elytron integration here will be a requirement - this will have the advantage that a SecurityIdentity is established on arriving at the service and will be in use for any ongoing invocation from that service.

Reply:
Ok, I totally defer to you and Farah, et al, regarding security.

Ah, so there's the scenario in which a JAX-RS resource makes a gRPC call. I didn't even think of that. My first reaction is that it would be yet another feature.